File: src/main/java/me/miki/shindo/ui/ClickEffects.java
package me.miki.shindo.ui;

import lombok.Getter;
import me.miki.shindo.Shindo;
import me.miki.shindo.management.color.AccentColor;
import me.miki.shindo.management.nanovg.NanoVGManager;
import me.miki.shindo.utils.ColorUtils;
import me.miki.shindo.utils.animation.normal.Animation;
import me.miki.shindo.utils.animation.normal.Direction;
import me.miki.shindo.utils.animation.normal.easing.EaseBackIn;

import java.util.ArrayList;

public class ClickEffects {

    @Getter
    private static ClickEffects instance;

    private final ArrayList<ClickEffect> effects = new ArrayList<ClickEffect>();
    private final ArrayList<ClickEffect> removeEffects = new ArrayList<ClickEffect>();

    public ClickEffects() {
        instance = this;
    }

    public void drawClickEffects() {

        for (ClickEffect ce : effects) {

            if (ce.isDone()) {
                removeEffects.add(ce);
            }

            ce.draw();
        }

        effects.removeAll(removeEffects);
    }

    public void addClickEffect(int mouseX, int mouseY) {
        effects.add(new ClickEffect(mouseX, mouseY));
    }

    private static class ClickEffect {

        private final Animation animation;
        private final int x;
        private final int y;

        private ClickEffect(int x, int y) {
            this.x = x;
            this.y = y;
            this.animation = new EaseBackIn(650, 1, 0.0F);
        }

        public void draw() {

            Shindo instance = Shindo.getInstance();
            NanoVGManager nvg = instance.getNanoVGManager();
            AccentColor currentColor = instance.getColorManager().getCurrentColor();

            nvg.setupAndDraw(() -> {
                nvg.drawArc(x, y, animation.getValueFloat() * 8, 0, 360, 2, ColorUtils.applyAlpha(currentColor.getInterpolateColor(0), (int) (255 - (animation.getValueFloat() * 255))));
            });
        }

        public boolean isDone() {
            return animation.isDone(Direction.FORWARDS);
        }
    }
}

File: src/main/kotlin/me/miki/shindo/ui/ClickEffects.kt
package me.miki.shindo.ui

import me.miki.shindo.Shindo
import me.miki.shindo.utils.ColorUtils
import me.miki.shindo.utils.animation.normal.Direction
import me.miki.shindo.utils.animation.normal.easing.EaseBackIn

class ClickEffects {
    private val effects = mutableListOf<ClickEffect>()
    private val toRemove = mutableListOf<ClickEffect>()

    fun drawClickEffects() {
        effects.forEach { effect ->
            if (effect.isDone()) {
                toRemove.add(effect)
            }
            effect.draw()
        }

        if (toRemove.isNotEmpty()) {
            effects.removeAll(toRemove)
            toRemove.clear()
        }
    }

    fun addClickEffect(mouseX: Int, mouseY: Int) {
        effects.add(ClickEffect(mouseX, mouseY))
    }

    private class ClickEffect(
        private val x: Int,
        private val y: Int
    ) {
        private val animation = EaseBackIn(650, 1.0, 0.0f)

        fun draw() {
            val instance = Shindo.getInstance()
            val nvg = instance.nanoVGManager ?: return
            val currentColor = instance.colorManager.currentColor

            nvg.setupAndDraw(Runnable {
                val progress = animation.getValueFloat()
                nvg.drawArc(
                    x.toFloat(),
                    y.toFloat(),
                    progress * 8,
                    0f,
                    360f,
                    2f,
                    ColorUtils.applyAlpha(currentColor.getInterpolateColor(0), (255 - (progress * 255)).toInt())
                )
            });
        }

        fun isDone(): Boolean = animation.isDone(Direction.FORWARDS)
    }
}